versioned/persist/nontx/src/main/java/org/projectnessie/versioned/persist/nontx/NonTransactionalDatabaseAdapter.java

```diff
@@ -832,63 +761,525 @@ protected Hash casOpLoop(
+                  .setVersion(randomHash().asBytes());
+          newRefLog.getParentsList().stream()
+              .limit(config.getParentsPerRefLogEntry() - 1)
+              .forEach(newRefLogParents::addRefLogParentsInclHead);
+
+          if (refLogParentsCas(ctx, stripe, refLogParents, newRefLogParents.build())) {
+            return tryState.success(result.newHead);
           }
-          Optional<Hash> newGlobalLogHead =
-              globalLogHead(pointer).equals(globalLogHead(newPointer))
-                  ? Optional.empty()
-                  : Optional.of(Hash.of(globalLogHead(newPointer)));
-          cleanUpCommitCas(
-              ctx,
-              newGlobalLogHead,
-              individualCommits,
-              individualKeyLists,
-              Hash.of(newPointer.getRefLogId()));
+
+          cleanUpRefLogWrite(ctx, Hash.of(newRefLog.getRefLogId()));
+
+          // Need to retry "forever" until the ref-log entry could be added.
+          tryState.retryEndless();
         }
+      }
+    }
+  }
 
-        tryState.retry();
+  protected static ReferenceInfo<ByteString> namedReferenceToReferenceInfo(NamedReference r) {
+    return ReferenceInfo.of(
+        Hash.of(r.getRef().getHash()), toNamedRef(r.getRef().getType(), r.getName()));
+  }
+
+  private final class RefLogSplit {
+    private final Spliterator<RefLog> source;
+
+    private boolean exhausted;
+
+    private RefLog current;
+
+    private RefLogSplit(NonTransactionalOperationContext ctx, int stripe) {
+      this.source = refLogStripeFetcher(ctx, stripe);
+    }
+
+    private void advance() {
+      if (exhausted) {
+        return;
+      }
+
+      current = null;
+      if (!source.tryAdvance(refLog -> current = refLog)) {
+        exhausted = true;
+      }
+    }
+
+    private void maybeAdvance() {
+      if (current == null) {
+        advance();
+      }
+    }
+
+    RefLog pull() {
+      if (exhausted) {
+        return null;
+      }
+
+      maybeAdvance();
+
+      RefLog r = current;
+      if (r != null) {
+        current = null;
+      }
+      return r;
+    }
+
+    long operationTime() {
+      maybeAdvance();
+
+      return current != null ? current.getOperationTime() : 0L;
+    }
+
+    boolean hasMore() {
+      maybeAdvance();
+
+      return current != null;
+    }
+
+    @Override
+    public String toString() {
+      return "RefLogSplit{" + "exhausted=" + exhausted + ", current=" + current + '}';
+    }
+  }
+
+  private final class RegLogSpliterator extends Spliterators.AbstractSpliterator<RefLog> {
+
+    private final List<RefLogSplit> splits;
+
+    private final Hash initialHash;
+    private boolean initialHashSeen;
+    private RefLog initialRefLog;
+
+    RegLogSpliterator(NonTransactionalOperationContext ctx, int stripes, Hash initialHash)
+        throws RefLogNotFoundException {
+      super(Long.MAX_VALUE, Spliterator.ORDERED);
+
+      this.initialHash = initialHash;
```

Comment:
It looks like `initialHash` has dual purpose - proper initial hash and pagination token. I suppose in the latter case we could use a composite token to keep track of cursors in all stripes.

Reply:
Hm - it's only used to identify the initial hash, there's no pagination.

Reply:
Oh *that* paging. Well, that's "external" paging, yea - that's a "bit" more inefficient at the moment.
Either leave it as it is in the current PR and "deprecate the reflog for removal" - or have a better "external" paging state, which isn't too hard either, but adds more code to this PR.

Reply:
With the goal of deprecating the reflog API soon, current impl. looks good enough for me.